### ISSUE TITLE
WebView2 crash OnWindowPositionChanged

### ISSUE THREAD

**Issue by raquibul-h** [MicrosoftEdge/WebView2Feedback#1531]
WebView2 crash OnWindowPositionChanged

Description
Exception Info: System.InvalidOperationException
at Microsoft.Web.WebView2.Wpf.WebView2.VerifyBrowserNotCrashed()
at Microsoft.Web.WebView2.Wpf.WebView2.get_CoreWebView2()
at Microsoft.Web.WebView2.Wpf.WebView2.OnWindowPositionChanged(System.Windows.Rect)
at System.Windows.Interop.HwndHost.UpdateWindowPos()
at System.Windows.Interop.HwndHost.OnLayoutUpdated(System.Object, System.EventArgs)
at System.Windows.ContextLayoutManager.fireLayoutUpdateEvent()
at System.Windows.ContextLayoutManager.UpdateLayout()
at System.Windows.ContextLayoutManager.UpdateLayoutCallback(System.Object)
at System.Windows.Media.MediaContext+InvokeOnRenderCallback.DoWork()
at System.Windows.Media.MediaContext.FireInvokeOnRenderCallbacks()
at System.Windows.Media.MediaContext.RenderMessageHandlerCore(System.Object)
at System.Windows.Media.MediaContext.RenderMessageHandler(System.Object)
at System.Windows.Threading.ExceptionWrapper.InternalRealCall(System.Delegate, System.Object, Int32)
at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(System.Object, System.Delegate, System.Object, Int32, System.Delegate)
at System.Windows.Threading.DispatcherOperation.InvokeImpl()
at System.Windows.Threading.DispatcherOperation.InvokeInSecurityContext(System.Object)
at MS.Internal.CulturePreservingExecutionContext.CallbackWrapper(System.Object)
at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
at MS.Internal.CulturePreservingExecutionContext.Run(MS.Internal.CulturePreservingExecutionContext, System.Threading.ContextCallback, System.Object)
at System.Windows.Threading.DispatcherOperation.Invoke()
at System.Windows.Threading.Dispatcher.ProcessQueue()
at System.Windows.Threading.Dispatcher.WndProcHook(IntPtr, Int32, IntPtr, IntPtr, Boolean ByRef)
at MS.Win32.HwndWrapper.WndProc(IntPtr, Int32, IntPtr, IntPtr, Boolean ByRef)
at MS.Win32.HwndSubclass.DispatcherCallbackOperation(System.Object)
at System.Windows.Threading.ExceptionWrapper.InternalRealCall(System.Delegate, System.Object, Int32)
at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(System.Object, System.Delegate, System.Object, Int32, System.Delegate)
at System.Windows.Threading.Dispatcher.LegacyInvokeImpl(System.Windows.Threading.DispatcherPriority, System.TimeSpan, System.Delegate, System.Object, Int32)
at MS.Win32.HwndSubclass.SubclassWndProc(IntPtr, Int32, IntPtr, IntPtr)
at MS.Win32.UnsafeNativeMethods.DispatchMessage(System.Windows.Interop.MSG ByRef)
at System.Windows.Threading.Dispatcher.PushFrameImpl(System.Windows.Threading.DispatcherFrame)
at System.Windows.Threading.Dispatcher.PushFrame(System.Windows.Threading.DispatcherFrame)
at System.Windows.Application.RunDispatcher(System.Object)
at System.Windows.Application.RunInternal(System.Windows.Window)
at System.Windows.Application.Run(System.Windows.Window)

Version
SDK: 1.0.865-prerelease
Runtime: 91.0.864.67
Framework: WPF
OS: Windows 10

Repro Steps
Window containing Webview2 object is tried to be shown. Webview2 crashes. Expected: Should not crash.

Screenshots

Additional context
Thanks for your response.
We are using Webview2 inside a WPF application which runs in the background with a system tray icon. Webview2 instance is alive throughout the WPF application lifecycle.

webview2 is initialized as: 
```cs
public void Init()
{
    this.webView = new WebView2();
    webView.Loaded += async (sender, e) =>
    {
        await webView.EnsureCoreWebView2Async();


        this.webView.NavigationCompleted += WebView_NavigationCompleted;
        this.webView.NavigationStarting += WebView_NavigationStarting;

        this.webView.CoreWebView2.ProcessFailed += CoreWebView_ProcessFailed;

        webView.Height = _height;
        webView.Width = _width;
        
        webView.VerticalAlignment = System.Windows.VerticalAlignment.Bottom;

        //these are dummy values.
        webView.CoreWebView2.SetVirtualHostNameToFolderMapping("hostName", _folderPath, CoreWebView2HostResourceAccessKind.Allow);
        webView.Source = new Uri(@"https://hostName/" + "index.html");
    };
}
```

The main window of the WPF application is shown on foreground on some user action. This main window has the webview in it.

```cs
private void Window_Loaded(object sender, RoutedEventArgs e)
{
     this.Content = this._webview;
}
```

On user action, we show this window using the .show() api.

We have also added the CoreWebView2.ProcessFailed event. Will share more details of the crash here.

Thanks!

(PS - I opened this as a duplicate of https://github.com/MicrosoftEdge/WebView2Feedback/issues/1518 since I couldn't comment)

@champnic - please check this issue instead. Thanks!

[AB#38523895](https://microsoft.visualstudio.com/90b2a23c-cab8-4e7c-90e7-a977f32c1f5d/_workitems/edit/38523895)

**Comment by champnic** [comment]
@raquibul-h Thanks for the extra info. Nothing is jumping out to me from the code. Do you hit this consistently every time? Does it also repro if you use an external website (like https://bing.com, for example)? Do you know if any of the initialization code in the Loaded handler is run by the time the crash happens?

**Comment by raquibul-h** [comment]
@champnic : it happens randomly (sometimes issue) and in no fixed pattern. The window which holds the webview2 object gets closed and reopened constantly. However for optimization, webview2 object is not destroyed. The object is owned by a singleton class. We simply set the content of the window to webview2 as shown above.

Haven't tried external website yet. Will check about the initialization code in loader.

**Comment by raquibul-h** [comment]
Also when this exception happens, whole app crashes. 

Can you tell suggest a way to prevent this?

**Comment by techyrajeev** [comment]
@champnic facing similar issue 

![image](https://user-images.githubusercontent.com/4118032/158164172-cf939bff-6453-4aea-a233-bdedb5ad4601.png)

**Comment by cbra-caa** [comment]
@champnic We get the following log statement when this happens:

`CoreWebView2_ProcessFailed - Core process failed (BrowserProcessExited - EmptyProcessDescription) with reason 'Unexpected' and exitcode '-805306369' - [WebView hash '16120740']`

Which is immediatly followed by:
```
OnDispatcherUnhandledException caught
Dispatcher: HasShutdownFinished=False, HasShutdownStarted=False
Dispatcher.Thread: ManagedThreadId=1, Name=Main, ThreadState=Running, IsAlive=True, IsBackground=False

HandleUnhandledException - Unhandled exception
System.InvalidOperationException: The WebView control is no longer valid because the browser process crashed.To work around this, please listen for the CoreWebView2.ProcessFailed event to explicitly manage the lifetime of the WebView2 control in the event of a browser failure.https://docs.microsoft.com/en-us/dotnet/api/microsoft.web.webview2.core.corewebview2.processfailed
   at Microsoft.Web.WebView2.Wpf.WebView2.VerifyBrowserNotCrashed()
   at Microsoft.Web.WebView2.Wpf.WebView2.get_CoreWebView2()
   at Microsoft.Web.WebView2.Wpf.WebView2.OnWindowPositionChanged(Rect rcBoundingBox)
   at System.Windows.Interop.HwndHost.UpdateWindowPos()
   at System.Windows.ContextLayoutManager.fireLayoutUpdateEvent()
   at System.Windows.ContextLayoutManager.UpdateLayout()
   at System.Windows.ContextLayoutManager.UpdateLayoutCallback(Object arg)
   at System.Windows.Media.MediaContext.FireInvokeOnRenderCallbacks()
   at System.Windows.Media.MediaContext.RenderMessageHandlerCore(Object resizedCompositionTarget)
   at System.Windows.Media.MediaContext.RenderMessageHandler(Object resizedCompositionTarget)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(Object source, Delegate callback, Object args, Int32 numArgs, Delegate catchHandler)
```

We catch the event, and begin disposing of the WebView on the ProcessFailed event, but it seems that this call is already on the stack and therefore gets executed regardless.

It originates from the user just activating a window with a running WebView inside.
`LogWindowActivated - Activated window`

**Comment by champnic** [comment]
@cbra-caa Ya that is problematic. I'll try and bump the priority on this one.

**Comment by techyrajeev** [comment]
Hi @champnic is there any fix available at this moment?
If not in which version we might expect fix to land soon?

**Comment by schroedl** [comment]
I added handlers for ProcessFailed and BrowserProcessExited in an attempt to prevent a browser fail from taking down our application. I can hit this exception every time if I run our app, open a browser window and then use Process Explorer to kill off the root msedgewebview.exe process. The ProcessFailed/BrowserProcessExited events do fire but there's nothing we can really do in there - what can we do? There's no e->Handled I could set. So I unhook events and try to disavow use of the browser control. I even tried creating a new instance of the browser to replace the failed one.  But, after the event handlers return we get this exact InvalidOperationException.

**Comment by champnic** [comment]
We've put in a fix so that we don't raise the exception in OnWindowPositionChanged, and it should be available in the next SDK prerelease package soon (this week or next). That exception wasn't helpful because you weren't left with any way to avoid or handle it gracefully.

**Comment by champnic** [comment]
This is available for testing in 1.0.1243-prerelease SDK. Thanks!
(Note: This SDK has been deprecated due to compat issues with the StatusBarTextChanged API, but should be safe to use if you don't use that API. See here for more info: https://github.com/MicrosoftEdge/WebView2Feedback/issues/2414#issuecomment-1116335991)

**Comment by jamesnet214** [comment]
WPF has the DefaultBackgroundColor property, so you can also change the default '#FFFFFF' to another color. This may not be the perfect solution, but it is a good alternative. 😄 
```
string hex = "#1a202f";
DefaultBackgroundColor = System.Drawing.ColorTranslator.FromHtml(hex);
```
